File: tests/libjournal/channel_index.cc
// -*- c++ -*-
//
// michael a.g. aïvázis <michael.aivazis@para-sim.com>
// (c) 1998-2020 all rights reserved


// get the journal
#include <p2/journal.h>
// support
#include <cassert>


// type aliases
template <bool stateV = true>
using inventory_t = pyre::journal::inventory_t<stateV>;

template <typename severityT, typename inventoryT>
using channel_t = pyre::journal::channel_t<severityT, inventoryT>;


// severity stub
class severity_t :
    public channel_t<severity_t, inventory_t<true>>
{
    // types
public:
    using channel_type = channel_t<severity_t, inventory_t<true>>;

    // metamethods
public:
    // index initialization is required...
    severity_t(const name_type &);
};

// stub implementation
severity_t::severity_t(const name_type & name) :
    channel_type(name)
{}


// exercise the channel state index
int main() {
    // make a couple of channels
    severity_t channel_1("test.channel_1");
    severity_t channel_2("test.channel_2");

    // get the index through the first channel
    const severity_t::index_type & index_1 = channel_1.index();
    // verify it has exactly two channels
    assert (index_1.size() == 2);
    // one of them is "test.channel1"
    assert (index_1.contains("test.channel_1"));
    // and the other is "test.channel2"
    assert (index_1.contains("test.channel_2"));

    // get the index through the second channel
    const severity_t::index_type & index_2 = channel_2.index();
    // verify it has exactly two channels
    assert (index_2.size() == 2);
    // one of them is "test.channel1"
    assert (index_2.contains("test.channel_1"));
    // and the other is "test.channel2"
    assert (index_2.contains("test.channel_2"));

    // all done
    return 0;
}


// end of file

File: tests/journal.pkg/channel_index.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-
#
# michael a.g. aïvázis <michael.aivazis@para-sim.com>
# (c) 1998-2020 all rights reserved


# externals
import collections  # for {defaultdict}


# channel state interface
def test():
    """
    Verify that all channel instances have access to the same index
    """
    # get the package
    from j2.Channel import Channel
    # the base class has no index, so make one
    index = collections.defaultdict(Channel.disabled_type)
    # and attach it
    Channel.index = index

    # make a couple of channels
    channel_1 = Channel(name="test.channel_1")
    channel_2 = Channel(name="test.channel_2")

    # get the index through the first channel
    index_1 = channel_1.index
    # verify it has exactly two channels
    assert len(index_1) == 2
    # one of them is the name of {channel_1}
    assert channel_1.name in index_1
    # and the other is the name of {channel_2}
    assert channel_2.name in index_1

    # repeat, through the second channel
    index_2 = channel_1.index
    # verify it has exactly two channels
    assert len(index_2) == 2
    # one of them is the name of {channel_1}
    assert channel_1.name in index_2
    # and the other is the name of {channel_2}
    assert channel_2.name in index_2

    # all done
    return


# main
if __name__ == "__main__":
    # run the test
    test()


# end of file
